### ISSUE TITLE
Version resolving for vcs hosted packages

### ISSUE THREAD

**Issue by akeeman** [python-poetry/poetry#3221]
- [x] I have searched the [issues](https://github.com/python-poetry/poetry/issues) of this repo and believe that this is not a duplicate.
- [x] I have searched the [documentation](https://python-poetry.org/docs/) and believe that my question is not covered.

## Feature Request

**TL;DR** I'd like to be able to resolve vcs hosted package versions just like PyPI hosted packages.

Consider the following package structure:
```
A (vcs hosted)
|-- B (vcs hosted)
|   `-- C ref=1.0.0 points to v1.0.0 (vcs hosted)
`-- C ref=1.1.0 points to v1.1.0 (vcs hosted)
```

If those packages would not be vcs hosted we could say in both `A` and `B`'s requirements that `C` should be at version `^1`, so we'd end up with the latest version, say `v1.1.0`. However, because these are hosted on a vcs, we currently end up with an error, because both references to `C` should, but don't, refer to the same ref.

My suggestion would be to be able to add a `version` property to vcs hosted packages, so we'd get for instance
```toml
C = {git = "ssh://git@github.com/someone/C.git", ref = "1.1.0", version = "^1"}
```
When the `version` property is given, we should treat this like a normal package, look at what version the package itself communicates, and use the latest compatible one. So in case of the example:

* B requires C ^1, and provides a candidate using a link to C ref 1.0.0, which is at version 1.0.0
* A requires C ^1, and provides a candidate using a link to C ref 1.1.0, which is at version 1.1.0

We could say we'll take C ref 1.1.0, because it provides the latest version.

In order to save the trouble and speed bump of cloning the repo to see what version it provides one also could require that the tag is equal to the version number, an that the ref is a ref to the tag, so in the example above, ref 1.0.0 and ref 1.1.0 are refs to tags that can directly be used as version numbers.

#### Using branches
An alternative that already works would be to create a branch `1` and use the following:
```toml
C = {git = "ssh://git@github.com/someone/C.git", branch = "1"}
```
This gives the overhead of maintaining an extra branch in a vcs, and it looses the control over the requirement of a version range smaller than the major version, e.g. `version = ">=1.1,<1.3"` is not a thing, as it simulates `version = "^1"`.
This could be overcome by adding a branch that does precisely that, but I think you the point by now.

**Comment by abn** [comment]
[Wearing personal hat, not mainter hat for the following comments.]

@akeeman this is something I have been pondering about recently. A few points to talk about:

- Would we say that we only support cases where branches/tags are semver pep440 comatible? ie. `([v]<semver> | <pep 440>)?
- Would the wheel metadata use the `ref` version? If we do that doesn't that invalidate the "resolved" version? Maybe it should be either `ref` or `version` and not both.

Implementation-wise, it might be sensible to treat "git dependencies" as being an index internally and adding a "Repository" interface  to them, so package versions can be discovered. However, I would think there are a few other issues around git dependencies that should be solved before we tackle this due to performance reasons.

**Comment by finswimmer** [comment]
Hello,

> TL;DR I'd like to be able to resolve vcs hosted package versions just like PyPI hosted packages.

I can see the benefit, but I'm not sure we should do this. IMO if someone points to a vcs (or a path dependency) they expect that exactly that package is installed, that is found in this location. If a vcs is now treated just like a private package repository, this expectation is not met anymore.

Furthermore I don't think this would work, if you build a package that then get's installed by another tool like pip. 

fin swimmer

**Comment by abn** [comment]
> Furthermore I don't think this would work, if you build a package that then get's installed by another tool like pip.

Wouldn't this be just a direct reference from the locked version? As it is today? I do not think this would change.

> IMO if someone points to a vcs (or a path dependency) they expect that exactly that package is installed, that is found in this location.

I do not expect this to change. The internal representation is what changes, but particularly only when resolving not when installing etc. 

>  If a vcs is now treated just like a private package repository, this expectation is not met anymore.

I am not sure I agree with this. All this changes is the ability to select the reference based on semver. I would suspect `ref` and `version` to be mutually exclusive here. All this would mean is that during resolution we select a compatible tag; the bit that I am not sure about is that right now a build does not require the lock file, but if this needs to work ethere should be either a fallback version (ie. ref) or build will need to require the lock file be present. Either way, I do not see this as departing from what we are already do.

**Comment by akeeman** [comment]
I mostly agree with @abn. Also I think that giving vcs refs multiple modes of operation (exact ref, and resolving versions) is a good thing.

> [finswimmer] Furthermore I don't think this would work, if you build a package that then get's installed by another tool like pip.

A PEP would help here. So that votes for discarding this PR.

The final points of @abn hit the spot where the pain is. Coming up with proposals I got to a different possible solution. So let me get back to what I posted initially wit a few alterations:

```
A (vcs hosted) <- application
|-- B (vcs hosted) <- library
|   |-- C ref=1.0.0 points to v1.0.0 (vcs hosted) <- library
|   `-- D ref=1.0.0 points to v1.0.0 (vcs hosted) <- library
`-- C ref=1.1.0 points to v1.1.0 (vcs hosted) <- library
```

I noticed that I'll have to include a reference to where `D` is hosted in the requirements of `A` for it to work, and that, given `B` is a library, the only reason to install `C` and `D` at `B` directly is when it's being developed. So what can be done is adding `C = "^1.0"` and `D = "^1.0"` to the dependencies of `B` and adding the git refs to the dev dependencies of `B`. The refs to a valid `B`, `C`, and `D` has to be added to `A` anyway.

So thinking of it this way reduces the hassle to
`B = {git = "...", ref = "1.0.0", version = "^1.0"}` = `B = "^1.0"`  + (`B = {git = "...", ref = "1.0.0"}` if `B` is the dependency root, else nothing)
